Improve robot hitboxes

kirillgla opened the pull request:
Also includes some minor fixes

kirillgla committed:
Fix compilation on Windows

kirillgla committed:
Fix robot radius to match specifications

kirillgla committed:
Remove obsolete call to friction initialization

kirillgla committed:
Make robot's shape match its icon and specifictions

EgorOrachyov commented:
А что именно не компилировалось на win?

kirillgla commented:
то, что мы обсуждали в чатике:
1. не резолвился M_PI
2. компилятор от студии, похоже, не может в синтаксис `{}`, как в том файле, что я поправил этим коммтиом. В коммите то, что посоветовал сделать Сультим

EgorOrachyov commented:
Как говорится, кого-то лечит, а кого-то калечит.
В c++ с такими вещами надо быть очень аккуратным

![image](https://user-images.githubusercontent.com/31863304/113900984-1326ba00-97d7-11eb-8daf-731f090d2e55.png)
![image](https://user-images.githubusercontent.com/31863304/113901553-a19b3b80-97d7-11eb-9d75-a4f3f16de0ee.png)

kirillgla commented:
что будем делать с этим?
не хотелось бы #ifdef win

EgorOrachyov commented:
`<corecrt_math_defines.h>` может это можно как-то заменить на стандарные `<math.h>` для С кода, и  `<cmath>`  для С++ кода. Если нужна только PI, то лучше эту константу в код руками добавить. 10 знаков после запятой вполне хватит

EgorOrachyov commented:
Опять же, на  win бывает много компиляторов, и не все поддерживают специфичные для MCVS дополнительные инклуд файлы, что дает еще +1 за то, чтобы на стандартные инклуды перейти

kirillgla committed:
Fix compilation outside of windows

EgorOrachyov commented on rfsim/sources/physics/PhysicsServer.cpp:
```diff
@@ -281,13 +281,37 @@ namespace rfsim {
+        const int circlePointCount = 6;
+        const int pointCount = frontPointCount + circlePointCount;
+        b2Vec2 points[pointCount];
+        // constants are for r = 0.085. Scale them for other robot sizes:
+        const double radiusInSpecs = 0.085;
+        const double frontOffset = 0.050f;
```
Как я понял, здесь эти величины в метрах.
Т.е. изменение размера робота = придется идти сюда и менять эти константы.
А можно ли перейти на % от радиуса, который как параметр уже передается (r)?

kirillgla replied:
можно, и это было бы достаточно разумно, но нам не платили за идеальный код)

kirillgla replied:
да, в метрах

kirillgla replied:
не ожидаю, что у нас будут разные роботы в разных сценариях

EgorOrachyov commented on rfsim/sources/physics/PhysicsServer.cpp:
```diff
@@ -281,13 +281,37 @@ namespace rfsim {
+        // constants are for r = 0.085. Scale them for other robot sizes:
+        const double radiusInSpecs = 0.085;
+        const double frontOffset = 0.050f;
+        double cornerY = sqrt(radiusInSpecs * radiusInSpecs - frontOffset * frontOffset);
+        double scale = r / radiusInSpecs;
+        points[0] = {(float) (frontOffset * scale), (float) (-cornerY * scale)};
+        points[1] = {(float)(frontOffset * scale), (float) (cornerY * scale)};
+        double initialPointAngle = acos(frontOffset / radiusInSpecs);
+        double backArc = 2 * M_PI - initialPointAngle * 2;
+        double stepAngle = backArc / circlePointCount; // What are you doing, step-angle?
+        for (int i = 0; i < circlePointCount; i += 1) {
+            double currentAngle = initialPointAngle + stepAngle * (i + 1);
+            double x = r * cos(currentAngle);
+            double y = r * sin(currentAngle);
```
Используемые функции sqrt, sin, cos, acos и тд. - это не плюсовые функции из cmath. Плюсовые функции доступны через пространство `std::<имя функции>` см. [ссылку](https://en.cppreference.com/w/cpp/numeric/math/sin). Надо переписать на std, тк С функции в С++ коде лучше не использвать без веских причин

kirillgla replied:
done

kirillgla committed:
Revert "Fix compilation outside of windows"

kirillgla committed:
Revert "Fix compilation on Windows"

kirillgla committed:
Use std functions

EgorOrachyov committed:
[Code] Fixes:
- New ball size
- Correct ball drawing
- Robot collision geom with % factor
- Math constans fixes
